File: frontend/src/api/timetable.ts
import apiClient from './client'

export interface TimetableTemplate {
  id: number
  school: string
  name: string
  periods: Array<{ name: string; start: string; end: string }>
}

export interface TimetableData {
  monday?: Array<{ period: string; course: string }>
  tuesday?: Array<{ period: string; course: string }>
  wednesday?: Array<{ period: string; course: string }>
  thursday?: Array<{ period: string; course: string }>
  friday?: Array<{ period: string; course: string }>
  saturday?: Array<{ period: string; course: string }>
  sunday?: Array<{ period: string; course: string }>
}

export interface Timetable {
  id: number
  user_id: string
  data: TimetableData
}

export interface FreeSlot {
  start: string
  end: string
}

export const timetableApi = {
  getTemplates: async (): Promise<TimetableTemplate[]> => {
    const response = await apiClient.get('/timetable/templates')
    return response.data
  },

  getTimetable: async (): Promise<Timetable> => {
    const response = await apiClient.get('/timetable')
    return response.data
  },

  saveTimetable: async (data: TimetableData): Promise<Timetable> => {
    const response = await apiClient.post('/timetable', { data })
    return response.data
  },

  getFreeSlots: async (weekday: string): Promise<{ weekday: string; slots: FreeSlot[] }> => {
    const response = await apiClient.get(`/timetable/free-slots?weekday=${weekday}`)
    return response.data
  },
}

File: backend/app/schemas/timetable.py
from pydantic import BaseModel
from typing import List, Dict, Optional
from datetime import time


class PeriodTemplate(BaseModel):
    name: str
    start: str  # "HH:MM"
    end: str


class TimetableTemplateResponse(BaseModel):
    id: int
    school: str
    name: str
    periods: List[PeriodTemplate]


class TimetableData(BaseModel):
    # 格式: { "monday": [{ "period": "1", "course": "Calculus" }, ...], ... }
    monday: Optional[List[Dict[str, str]]] = None
    tuesday: Optional[List[Dict[str, str]]] = None
    wednesday: Optional[List[Dict[str, str]]] = None
    thursday: Optional[List[Dict[str, str]]] = None
    friday: Optional[List[Dict[str, str]]] = None
    saturday: Optional[List[Dict[str, str]]] = None
    sunday: Optional[List[Dict[str, str]]] = None


class TimetableCreate(BaseModel):
    data: TimetableData


class TimetableResponse(BaseModel):
    id: int
    user_id: str
    data: TimetableData


class FreeSlot(BaseModel):
    start: str  # "HH:MM"
    end: str


class FreeSlotsResponse(BaseModel):
    weekday: str
    slots: List[FreeSlot]

